launcher/FileSystem.cpp

```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+        } else {
+            if (m_debug)
+                qDebug() << "making symlink:" << src_path << "to" << dst_path;
+            fs::create_symlink(StringUtils::toStdString(src_path), StringUtils::toStdString(dst_path), m_os_err);
+        }
+
+        if (m_os_err) {
+            qWarning() << "Failed to link files:" << QString::fromStdString(m_os_err.message());
+            qDebug() << "Source file:" << src_path;
+            qDebug() << "Destination file:" << dst_path;
+            qDebug() << "Error category:" << m_os_err.category().name();
+            qDebug() << "Error code:" << m_os_err.value();
+            emit linkFailed(src_path, dst_path, QString::fromStdString(m_os_err.message()), m_os_err.value());
+        } else {
+            m_linked++;
+            emit fileLinked(src_path, dst_path);
+        }
+        if (m_os_err)
+            return false;
+    }
+    return true;
+}
+
+void create_link::runPrivileged(const QString& offset)
+{
+    m_linked = 0;  // reset counter
+    m_path_results.clear();
+    m_links_to_make.clear();
+
+    bool gotResults = false;
+
+    make_link_list(offset);
+
+    QString serverName = BuildConfig.LAUNCHER_APP_BINARY_NAME + "_filelink_server" + StringUtils::getRandomAlphaNumeric();
+
+    connect(&m_linkServer, &QLocalServer::newConnection, this, [&]() {
+        qDebug() << "Client connected, sending out pairs";
+        // construct block of data to send
+        QByteArray block;
+        QDataStream out(&block, QIODevice::WriteOnly);
+        out.setVersion(QDataStream::Qt_5_0);  // choose correct version better?
+
+        qint32 blocksize = quint32(sizeof(quint32));
+        for (auto link : m_links_to_make) {
+            blocksize += quint32(link.src.size());
+            blocksize += quint32(link.dst.size());
+        }
+        qDebug() << "About to write block of size:" << blocksize;
+        out << blocksize;
+
+        out << quint32(m_links_to_make.length());
+        for (auto link : m_links_to_make) {
+            out << link.src;
+            out << link.dst;
+        }
+
+        QLocalSocket* clientConnection = m_linkServer.nextPendingConnection();
+        connect(clientConnection, &QLocalSocket::disconnected, clientConnection, &QLocalSocket::deleteLater);
+
+        connect(clientConnection, &QLocalSocket::readyRead, this, [&, clientConnection]() {
+            QDataStream in;
+            quint32 blockSize = 0;
+            in.setDevice(clientConnection);
+            in.setVersion(QDataStream::Qt_5_0);
+            qDebug() << "Reading path results from client";
+            qDebug() << "bytes available" << clientConnection->bytesAvailable();
+
+            // Relies on the fact that QDataStream serializes a quint32 into
+            // sizeof(quint32) bytes
+            if (clientConnection->bytesAvailable() < (int)sizeof(quint32))
+                return;
```

Comment:
it is theoretically possible for a packet of data from one client to not arrive complete and we need to wait for the rest of the bytes before attempting to decode the data form the stream.
In practice at the level of local sockets something would need to go horridly wrong or someone would need to be trying something jank like proxying the socket. Still it's a good idea to have this here just in case.